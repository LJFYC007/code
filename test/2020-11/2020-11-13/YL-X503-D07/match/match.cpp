/***************************************************************
	File name: match.cpp
	Author: ljfcnyali
	Create time: 2020年11月13日 星期五 08时16分49秒
***************************************************************/
#include<bits/stdc++.h>

using namespace std;

#define int unsigned long long
#define REP(i, a, b) for ( int i = (a), _end_ = (b); i <= _end_; ++ i ) 
#define mem(a) memset ( (a), 0, sizeof ( a ) ) 
#define str(a) strlen ( a ) 
#define pii pair<int, int>
typedef long long LL;

const int maxn = 1e6 + 10;

LL S, P, T, n;
int a[maxn], tot;
__int128 t[maxn];
pii Stack[maxn];

int rnd() 
{
    S^=S<<13,S^=S>>17,S^=S<<5;
    return (S%P+P)%P;
}

inline int get(__int128 l, __int128 r) { return (l + r) * (r - l + 1) / 2; }

signed main()
{
    freopen("match.in", "r", stdin);
    freopen("match.out", "w", stdout);
    scanf("%lld%lld", &S, &P);
    scanf("%lld", &T); 
    while ( T -- ) 
    {
        scanf("%lld", &n); 
        REP(i, 1, n) a[i] = rnd();
        mem(t); tot = 1; Stack[1] = pii(0, 0); a[n + 1] = 0;
    
        REP(i, 1, n + 1)
        {
            int lst = i;
            while ( Stack[tot].first > a[i] )  
            {
                t[i - Stack[tot].second] += Stack[tot].first - max(a[i], Stack[tot - 1].first);
                lst = Stack[tot].second; -- tot;
            }
            if ( Stack[tot].first < a[i] ) Stack[++ tot] = pii(a[i], lst);
        }
        int ans = 0; __int128 now = 0;
        for ( int i = n; i >= 1; -- i ) 
        {
            ans += get(now + 1, now + t[i]) * i;
            now += t[i];
        }
        printf("%llu\n", ans);
    }
    return 0;
}
