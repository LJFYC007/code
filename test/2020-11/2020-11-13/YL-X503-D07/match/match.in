326453276652 1000000007
30
1000000
5000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
1000000
