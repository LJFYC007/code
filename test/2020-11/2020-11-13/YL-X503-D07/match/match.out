6257362517594013314
10745364984604095562
14600491869273615914
6458845049931557037
3123722872206660063
7993213705643351998
17695289963131191726
12115963843868516280
12259055905846814736
3102146494636668188
15105521397732622198
4266339287382122803
12228312688148172254
6097729336528316777
11904975870836636468
4026654019111584320
8958020655186463816
8063497366843945223
4017251801686965503
6118557042085890823
14576113490554760206
18425134814242873159
7893032423053347427
4024011834065884042
11124138885215357932
16399655970576105764
1303340431644916592
10194335639301601795
7293214890293309304
6890924901210019621
